---
title: "iPSC : Batch effect modeling on all genes"
author: "Kushal K Dey"
date: "2015-09-03"
output: 
  html_document:
    toc: true
---

## Objective

In this script, we use a batch effect modeling of the single cells using only the cell cycle genes and then see if on removing the batch effects and the individual effects, the residuals indeed cluster by the cell phase or cell cycle or not. 


We first load the packages and the data.

```{r packages, echo=FALSE, eval=TRUE, message=FALSE, warning=FALSE}
library(CountClust)
library(data.table)
library(maptpx)
library(gplots)
library(philentropy)
library(dplyr)
library(edgeR)
library(qtlcharts)
library(limma)
```

## Preparing the Data 

```{r}
reads <- data.frame(fread('../data/reads.txt'),row.names=1);
molecules <-  data.frame(fread('../data/molecules.txt'),row.names=1);
```

```{r}
quality_single_cells <- scan("../data/quality-single-cells.txt",
                             what = "character")
anno <- data.frame(fread('../data/annotation.txt'));
```

```{r}
molecules <- molecules[, grepl("bulk", colnames(molecules)) |
                         colnames(molecules) %in% quality_single_cells]
anno <- anno[anno$well == "bulk" | anno$sample_id %in% quality_single_cells, ]
stopifnot(ncol(molecules) == nrow(anno),
          colnames(molecules) == anno$sample_id)

reads <- reads[, grepl("bulk", colnames(reads)) |
                         colnames(reads) %in% quality_single_cells]
stopifnot(ncol(reads) == nrow(anno),
          colnames(reads) == anno$sample_id)
```

```{r}
expressed <- rowSums(molecules[, anno$well == "bulk"]) > 0 &
             rowSums(molecules[, anno$well != "bulk"]) > 0
molecules <- molecules[expressed, ]

expressed <- rowSums(reads[, anno$well == "bulk"]) > 0 &
             rowSums(reads[, anno$well != "bulk"]) > 0
reads <- reads[expressed, ]

molecules_single <- molecules %>% select(-contains("bulk"))
reads_single <- reads %>% select(-contains("bulk"))

## Remove ERCC genes and genes with max molecule numer larger than 1024

ii_nonERCC <- grep("ERCC", rownames(molecules_single), invert = TRUE)

molecules_single <- molecules_single[ii_nonERCC,];

ii_notOverexpressed <- apply(molecules_single, 2, function(x) any(x < 1024) )

molecules_single <- molecules_single[ii_notOverexpressed,];

dim(molecules_single)


```


Now we draw a list of marker genes that have cell cycle information.

```{r}

cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})

labs=unique(unlist(lapply(1:5, function(k) match(cell_cycle_genes_list[[k]],rownames(molecules_single)))));
labs = labs[!is.na(labs)];
molecules_single_cell_cycle <- molecules_single[labs,];
```



We fix the cell phase scores

```{r phase_scores, echo=TRUE, eval=TRUE}
ans <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
  #### add average expression of all genes in the phase
  combined_matrix <- rbind(reads_single_phase,average=apply(reads_single_phase,2,mean))
  #### use transpose to compute cor matrix
  cor_matrix <- cor(t(combined_matrix))
  #### take the numbers
  cor_vector <- cor_matrix[,dim(cor_matrix)[1]]
  #### restrict to correlation >= 0.3 
  reads_single_phase_restricted <- reads_single_phase[rownames(reads_single_phase) %in% names(cor_vector[cor_vector >= 0.3]),]
  #### apply normalization to reads
  norm_factors_single <- calcNormFactors(reads_single_phase_restricted, method = "TMM")
  reads_single_cpm <- cpm(reads_single_phase_restricted, log = TRUE,
                            lib.size = colSums(reads_single) * norm_factors_single)
  #### output the phase specific scores (mean of normalized expression levels in the phase)
  apply(reads_single_cpm,2,mean)

})

#### normalization function
flexible_normalization <- function(data_in,by_row=TRUE){
  if(by_row){
    row_mean <- apply(data_in,1,mean)
    row_sd   <- apply(data_in,1,sd)
    output <- data_in
    for(i in 1:dim(data_in)[1]){
      output[i,] <- (data_in[i,] - row_mean[i])/row_sd[i]
    }
  }
  #### if by column
  if(!by_row){
    col_mean <- apply(data_in,2,mean)
    col_sd   <- apply(data_in,2,sd)
    output <- data_in
    for(i in 1:dim(data_in)[2]){
      output[,i] <- (data_in[,i] - col_mean[i])/col_sd[i]
    }
  }
  output
}

#### apply the normalization function
## first normalized for each phase
ans_normed <- flexible_normalization(ans,by_row=FALSE)
## then normalized of each cell
ans_normed_normed <- flexible_normalization(ans_normed,by_row=TRUE)

cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])
assign_cell_phase <- data.frame(cell_phase)
cell_phase_vector <- as.vector(as.matrix(assign_cell_phase));
cell_phase_vector <- factor(cell_phase_vector, 
                            levels = c("G1.S", "S", "G2.M", "M", "M.G1"))
```


Now we apply a linear model taking into account the individual effect and the batch effect and grab the residuals free of individual and batch effects.

```{r echo=TRUE, eval=TRUE}

counts_data <- t(molecules_single);

individual_id <- sapply(1:length(rownames(counts_data)), function(x) strsplit(rownames(counts_data)[x],"[.]")[[1]][1]);

batch_id <- sapply(1:length(rownames(counts_data)), function(x) strsplit(rownames(counts_data)[x],"[.]")[[1]][2]);

individual.batch.id <- paste0(individual_id,"_", batch_id);
batch_corrected_counts <- BatchCorrectedCounts(counts_data,individual.batch.id);

#write.table(batch_corrected_counts,"../data/batch_removed_counts_all_genes.txt");

nclus_vec <- 2:5
samp_metadata <- cbind.data.frame(individual_id, cell_phase_vector)
colnames(samp_metadata) <- c("individual", "cell_phase")

if(!dir.exists("../figures/ipsc_structure")) dir.create("../figures/ipsc_structure")

if(file.exists("../../project/rdas/topic_fit_ipsc.rda")) {
ipsc_topics_single <- get(load("../../project/rdas/topic_fit_ipsc.rda"));
} else {
StructureObj(counts_data ,nclus_vec,samp_metadata = samp_metadata, tol=0.005, batch_lab = batch_id, path_rda="../../project/rdas/topic_fit_ipsc.rda",path_struct = "../figures/ipsc_structure");
  
ipsc_topics_single <- get(load("../../project/rdas/topic_fit_ipsc.rda"));
}

for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(ipsc_topics_single[[num]]$omega, samp_metadata = samp_metadata, batch_lab = batch_id,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct='../figures/ipsc_structure',control=list(cex.axis=1));
}

```

## Batch Uncorrected Structure Plot

### Structure plot by individual

<img src='../figures/ipsc_structure/clus_2/struct_clus_2_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_3/struct_clus_3_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_4/struct_clus_4_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_5/struct_clus_5_individual.png' style="width:304px;height:228px;">

### Structure plot by batch

<img src='../figures/ipsc_structure/clus_2/struct_clus_2_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_3/struct_clus_3_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_4/struct_clus_4_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_5/struct_clus_5_batch.png' style="width:304px;height:228px;">

### Structure plot by cell phase

<img src='../figures/ipsc_structure/clus_2/struct_clus_2_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_3/struct_clus_3_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_4/struct_clus_4_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure/clus_5/struct_clus_5_cell_phase.png' style="width:304px;height:228px;">


```{r }

if(!dir.exists("../figures/ipsc_structure_batchcorrect")) dir.create("../figures/ipsc_structure_batchcorrect")

if(file.exists("../../project/rdas/topic_fit_ipsc_batchcorrect.rda")) {
ipsc_topics_single_batchcorrect <- get(load("../../project/rdas/topic_fit_ipsc_batchcorrect.rda"));
} else {
StructureObj(batch_corrected_counts ,nclus_vec,samp_metadata = samp_metadata, tol=0.005, batch_lab = batch_id, path_rda="../../project/rdas/topic_fit_ipsc_batchcorrect.rda",path_struct = "../figures/ipsc_structure_batchcorrect");
  
ipsc_topics_single_batchcorrect <- get(load("../../project/rdas/topic_fit_ipsc_batchcorrect.rda"));
}

for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(ipsc_topics_single_batchcorrect[[num]]$omega, samp_metadata = samp_metadata, batch_lab = batch_id,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct='../figures/ipsc_structure_batchcorrect',control=list(cex.axis=1));
}

```

## Batch Corrected Structure Plot

### Structure plot by individual

<img src='../figures/ipsc_structure_batchcorrect/clus_2/struct_clus_2_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_3/struct_clus_3_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_4/struct_clus_4_individual.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_5/struct_clus_5_individual.png' style="width:304px;height:228px;">

### Structure plot by batch

<img src='../figures/ipsc_structure_batchcorrect/clus_2/struct_clus_2_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_3/struct_clus_3_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_4/struct_clus_4_batch.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_5/struct_clus_5_batch.png' style="width:304px;height:228px;">

### Structure plot by cell phase

<img src='../figures/ipsc_structure_batchcorrect/clus_2/struct_clus_2_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_3/struct_clus_3_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_4/struct_clus_4_cell_phase.png' style="width:304px;height:228px;">
<img src='../figures/ipsc_structure_batchcorrect/clus_5/struct_clus_5_cell_phase.png' style="width:304px;height:228px;">
