---
title: 'Buettner single cell data (with single cell training): classtpx check'
author: "Kushal K Dey"
date: "March 17, 2016"
output: 
  html_document:
    css: floating-toc.css
    toc: true
---


```{r echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
#rm(list=ls())
library(maptpx)
library(CountClust)
library(classtpx)
library(limma)
library(data.table)
library(e1071)
```


## Overview

When `classtpx` was used on the combined single cell RNA-seq FACS sorted data due to [Buettner et al 2015](http://www.nature.com/nbt/journal/v33/n2/full/nbt.3102.html) and the bulk RNA FACS sorted data due to [Scialdone et al 2015](http://www.sciencedirect.com/science/article/pii/S1046202315300098). The results were not very satisfactory as the classifier on test data would always put them in the G1 phase cluster. 

The idea now is to explore the Buettner et al data and try to see if the difference in library size between the single cell RNA-seq and the bulk RNA-seq model when used jointly as in this case, is indeed a problem. 

## Buettner et al 2015 data 

```{r echo=TRUE, eval=TRUE}
library(data.table)
G1_single <- data.frame(fread('../data/Marioni_data/G1_singlecells_counts.txt'), row.names=1);
G2M_single <- data.frame(fread('../data/Marioni_data/G2M_singlecells_counts.txt'), row.names=1);
S_single <- data.frame(fread('../data/Marioni_data/S_singlecells_counts.txt'), row.names=1);

cell_phases <- c(rep("G1", 96), rep("S", 96), rep("G2M", 96))
```

We filter out ERCC spike ins.

```{r echo=TRUE, eval=TRUE}
ercc_start <- grep("ERCC", rownames(G1_single))[1]
G1_single <- G1_single[-(ercc_start:dim(G1_single)[1]),-(1:3)];
G2M_single <- G2M_single[-(ercc_start:dim(G2M_single)[1]),-(1:3)];
S_single <- S_single[-(ercc_start:dim(S_single)[1]),-(1:3)];
pooled_data <- t(cbind(G1_single, S_single, G2M_single));
```

## classtpx modeling

```{r echo=TRUE, eval=TRUE}
class_labs_full <- rep(1:3, each=96)
known_samples <- c(1:48, 97:144, 193:240);
class_labs <- rep(1:3, each=48)
```

```{r}
training_data <- pooled_data[known_samples,];
```

```{r}
mean_features <- apply(training_data, 2, function(x) return (tapply(x, as.factor(class_labs), function(y) return(round(mean(y))))));
mean_features <- class.normalizetpx(mean_features, byrow=TRUE)
top_features <- CountClust::ExtractTopFeatures(t(mean_features), top_features = 2000,                                             method="poisson", options="min")
top_features <- as.vector(top_features)
top_features <- top_features[!is.na(top_features)]
```


```{r}
top_features <- 1:dim(pooled_data)[2]
train.x <- pooled_data[known_samples,top_features];
train.y <- class_labs;
test.x <- pooled_data[-(known_samples),top_features];
test.y <- class_labs_full[-(known_samples)];
```


```{r}
#######  PLS-LDA on Buettner V6 counts #########

out2 <- plsgenomics::pls.lda(train.x, train.y, test.x, ncomp=50)
test_class_pls <- out2$predclass
tab_class_pls <- table(test_class_pls, test.y)
misclass_pls <- sum(tab_class_pls[row(tab_class_pls)!=col(tab_class_pls)])/ sum(tab_class_pls)
misclass_pls

#######  PLS-LDA on Buettner V6 voom #########

out2 <- plsgenomics::pls.lda(voom2(train.x), train.y, voom2(test.x), ncomp=50)
test_class_pls <- out2$predclass
tab_class_pls <- table(test_class_pls, test.y)
misclass_pls <- sum(tab_class_pls[row(tab_class_pls)!=col(tab_class_pls)])/ sum(tab_class_pls)
misclass_pls

#######  PLS-LDA on Buettner voom shrink #########

voom_generate <- class.voom_generator(train.x, class_labs = class_labs, doshrink=TRUE)

out2 <- plsgenomics::pls.lda(voom_generate$voom_shrunk_class, train.y, voom2(test.x), ncomp=50)
test_class_pls <- out2$predclass
tab_class_pls <- table(test_class_pls, test.y)
misclass_pls <- sum(tab_class_pls[row(tab_class_pls)!=col(tab_class_pls)])/ sum(tab_class_pls)
misclass_pls

```


```{r}

voom_generate <- class.voom_generator(train.x, class_labs = class_labs, doshrink=TRUE)


#############   SVM on Buettner counts data ###########################

training.data.frame <- data.frame(cbind(train.y, train.x));
colnames(training.data.frame)[-1] <- colnames(train.x)
model <- svm(as.factor(train.y) ~ ., data = training.data.frame)
colnames(test.x) <- colnames(train.x)
test_class_svm <- predict(model, test.x)
tab_class_svm <- table(test_class_svm, test.y)
misclass_svm <- sum(tab_class_svm[row(tab_class_svm)!=col(tab_class_svm)])/ sum(tab_class_svm)
misclass_svm

#############  SVM on Buettner voom data ##############################

voom.train.x <- voom2(train.x)
voom.test.x <- voom2(test.x)
training.data.frame <- data.frame(train.y, voom.train.x);
colnames(training.data.frame)[-1] <- colnames(train.x)
model <- svm(as.factor(train.y) ~ ., data = training.data.frame)
colnames(voom.test.x) <- colnames(voom.train.x)
test_class_svm <- predict(model, voom.test.x)
tab_class_svm <- table(test_class_svm, test.y)
misclass_svm <- sum(tab_class_svm[row(tab_class_svm)!=col(tab_class_svm)])/ sum(tab_class_svm)
misclass_svm

############  SVM on Buettner voom shrunk data ##########################

voom.train.x <- data.frame(voom_generate$voom_shrunk_class);
colnames(voom.train.x) <- colnames(train.x)
voom.test.x <- voom2(test.x)
training.data.frame <- data.frame(train.y, voom.train.x);
colnames(training.data.frame)[-1] <- colnames(train.x)
model <- svm(as.factor(train.y) ~ ., data = training.data.frame)
colnames(voom.test.x) <- colnames(voom.train.x)
test_class_svm <- predict(model, voom.test.x)
tab_class_svm <- table(test_class_svm, test.y)
misclass_svm <- sum(tab_class_svm[row(tab_class_svm)!=col(tab_class_svm)])/ sum(tab_class_svm)
misclass_svm

```

```{r}
counts <- pooled_data
model1 <- class.model_clust(counts, known_samples = known_samples,class_labs = class_labs, dist="normal")

test_class_normal <- apply(model1, 1, which.max);
tab_class_normal <- table(test_class_normal, test.y)

test_class_normal <- apply(model1, 1, which.max);
tab_class_normal <- table(test_class_normal, test.y)
print(tab_class_normal)
sum(diag(tab_class_normal[rownames(tab_class_normal) %in% colnames(tab_class_normal), ]))
misclass_normal <- 1 - (sum(diag(tab_class_normal[rownames(tab_class_normal) %in% colnames(tab_class_normal), ]))/ sum(tab_class_normal))
misclass_normal

print(tab_class_normal)



model2 <- class.model_clust(counts, known_samples = known_samples,
                            class_labs = class_labs, dist="poisson")

test_class_poisson <- apply(model2, 1, which.max);
tab_class_poisson <- table(test_class_poisson, test.y)
print(tab_class_poisson)
sum(diag(tab_class_poisson[rownames(tab_class_poisson) %in% colnames(tab_class_poisson), ]))
misclass_poisson <- 1 - (sum(diag(tab_class_poisson[rownames(tab_class_poisson) %in% colnames(tab_class_poisson), ]))/ sum(tab_class_poisson))
misclass_poisson

print(tab_class_poisson)


```

### omega.fix method

```{r echo=TRUE, eval=FALSE}
Topic_clus <- class_topics(
    counts, 
    K=3, 
    known_samples = known_samples,
    class_labs = class_labs,
    method="omega.fix",
    shrink=FALSE,
    shrink.method = 1,
    tol=0.01,
    ord=FALSE)


saveRDS(Topic_clus, file="../rdas/buettner_topic_fit_classtpx_omega_fix_singlecell.rda")
```

```{r echo=TRUE, eval=TRUE}
Topic_clus <- readRDS("../rdas/buettner_topic_fit_classtpx_omega_fix_singlecell.rda")

omega <- Topic_clus$omega;

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = c(rep("G1", 96), rep("S", 96), rep("G2", 96))
)

rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))

test_class_classtpx <- apply(omega,1,function(x) which.max(x));
tab_class_classtpx <- table(test_class_classtpx[-known_samples], test.y)
misclass_classtpx <- sum(tab_class_classtpx[row(tab_class_classtpx)!=col(tab_class_classtpx)])/ sum(tab_class_classtpx)
misclass_classtpx
```



### theta.fix method

```{r echo=TRUE, eval=FALSE}
Topic_clus <- class_topics(
    counts, 
    K=3, 
    known_samples = known_samples,
    class_labs = class_labs,
    method="theta.fix",
    shrink=FALSE,
    shrink.method = 1,
    tol=0.001,
    ord=FALSE)

saveRDS(Topic_clus, "../rdas/buettner_topic_fit_classtpx_theta_fix_singlecell_noshrink.rda")
```

```{r echo=TRUE, eval=TRUE}
Topic_clus <- readRDS("../rdas/buettner_topic_fit_classtpx_theta_fix_singlecell_noshrink.rda")

omega <- Topic_clus$omega;

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = c(rep("G1", 96), rep("S", 96), rep("G2", 96))
)

rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))

test_class_classtpx <- apply(omega,1,function(x) which.max(x));
tab_class_classtpx <- table(test_class_classtpx[-known_samples], test.y)
misclass_classtpx <- sum(tab_class_classtpx[row(tab_class_classtpx)!=col(tab_class_classtpx)])/ sum(tab_class_classtpx)
misclass_classtpx

```

### theta.fix + ash method

```{r echo=TRUE, eval=FALSE}
Topic_clus <- class_topics(
    counts, 
    K=3, 
    known_samples = known_samples,
    class_labs = class_labs,
    method="theta.fix",
    shrink=TRUE,
    shrink.method = 1,
    tol=0.001,
    ord=FALSE)

saveRDS(Topic_clus, "../rdas/buettner_topic_fit_classtpx_theta_prior_singlecell_shrink.rda")
```

```{r echo=TRUE, eval=TRUE}
Topic_clus <- readRDS("../rdas/buettner_topic_fit_classtpx_theta_prior_singlecell_shrink.rda")

omega <- Topic_clus$omega;

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = c(rep("G1", 96), rep("S", 96), rep("G2", 96))
)

rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))

test_class_classtpx <- apply(omega,1,function(x) which.max(x));
tab_class_classtpx <- table(test_class_classtpx[-known_samples], test.y)
misclass_classtpx <- sum(tab_class_classtpx[row(tab_class_classtpx)!=col(tab_class_classtpx)])/ sum(tab_class_classtpx)
misclass_classtpx

```


## maptpx Modeling

```{r echo=TRUE, eval=FALSE}
Topic_clus <- maptpx::topics(pooled_data, K=3, tol=0.01);
saveRDS(Topic_clus, "../rdas/buettner_topic_fit_maptpx.rda")
```

```{r echo=TRUE, eval=TRUE}
Topic_clus <- readRDS("../rdas/buettner_topic_fit_maptpx.rda")

omega <- Topic_clus$omega;

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = c(rep("G1", 96), rep("S", 96), rep("G2", 96))
)

rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))

```



## Conclusions

`classtpx` does a good job at separating the Buettner single cells into their cell cycle phases under omega.fix method compared to the theta.prior or theta.fix methods. The theta.fix method is better suited than the theta.prior method.
