---
title: "Clustering of transcriptional profile based on cell cycle genes"
author: "Kushal K Dey"
date: "September 3, 2015"
output: html_document
---

## Goal

Consider only the transcriptional profile of the cell-cycle genes. We apply an admixture model to the single cell data collected at Yoav's lab.  tSNE model +PCA model approach on the single cell data collected at Yoav's lab, using only the cell cycle gene expression. 

```{r echo=TRUE, eval=TRUE}
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(maptpx)))
suppressMessages(suppressWarnings(library(gplots)))
suppressMessages(suppressWarnings(library(philentropy)))
suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(edgeR)))
suppressMessages(suppressWarnings(library(tsne)))
suppressMessages(suppressWarnings(library(qtlcharts)))
```


## Prepare data

Import raw counts.

```{r echo=FALSE, eval=TRUE}
reads <- data.frame(fread('./data/reads.txt'),row.names=1);
molecules <-  data.frame(fread('./data/molecules.txt'),row.names=1);
```

Import quality single cell index.

```{r echo=FALSE, eval=TRUE}
quality_single_cells <- scan("./data/quality-single-cells.txt",
                             what = "character")
anno <- data.frame(fread('./data/annotation.txt'));
```

Keep only quality single cells. Remove bulk RNA-seq samples.

```{r, echo=FALSE, eval=TRUE}
molecules <- molecules[, grepl("bulk", colnames(molecules)) |
                         colnames(molecules) %in% quality_single_cells]
anno <- anno[anno$well == "bulk" | anno$sample_id %in% quality_single_cells, ]
stopifnot(ncol(molecules) == nrow(anno),
          colnames(molecules) == anno$sample_id)

reads <- reads[, grepl("bulk", colnames(reads)) |
                         colnames(reads) %in% quality_single_cells]
stopifnot(ncol(reads) == nrow(anno),
          colnames(reads) == anno$sample_id)
```

Exclude genes with no mapped reads across the single cell samples.

```{r, echo=FALSE, eval=TRUE}
expressed <- rowSums(molecules[, anno$well == "bulk"]) > 0 &
             rowSums(molecules[, anno$well != "bulk"]) > 0
molecules <- molecules[expressed, ]

expressed <- rowSums(reads[, anno$well == "bulk"]) > 0 &
             rowSums(reads[, anno$well != "bulk"]) > 0
reads <- reads[expressed, ]


molecules_single <- molecules %>% select(-contains("bulk"))
reads_single <- reads %>% select(-contains("bulk"))
```


Remove genes with max molecule numer larger than 1024

```{r}
molecules_single <- molecules_single[apply(molecules_single,1,max) < 1024,];
```



Now we import the list of marker gene for cell cycle information.

```{r echo=TRUE, eval=TRUE}
cell_cycle_genes <- read.table("./data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})
```

We extract those columns (genes) that match with the cell-cycle genes list.

```{r echo=TRUE, eval=TRUE}
labs <- unique(unlist(lapply(1:5, function(k) X = 
                                 match(cell_cycle_genes_list[[k]],
                                       rownames(molecules_single)))) )
labs <-  labs[!is.na(labs)]
molecules_single_cell_cycle <- molecules_single[labs,]
```


We have now extracted the data for the cell cycle genes only. Based on these cell cycle genes, we assign each cell to one of the phase scores. 

```{r phase_scores, echo=TRUE, eval=TRUE}
ans <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
  #### add average expression of all genes in the phase
  combined_matrix <- rbind(reads_single_phase,average=apply(reads_single_phase,2,mean))
  #### use transpose to compute cor matrix
  cor_matrix <- cor(t(combined_matrix))
  #### take the numbers
  cor_vector <- cor_matrix[,dim(cor_matrix)[1]]
  #### restrict to correlation >= 0.3 
  reads_single_phase_restricted <- reads_single_phase[rownames(reads_single_phase) %in% names(cor_vector[cor_vector >= 0.3]),]
  #### apply normalization to reads
  norm_factors_single <- calcNormFactors(reads_single_phase_restricted, method = "TMM")
  reads_single_cpm <- cpm(reads_single_phase_restricted, log = TRUE,
                            lib.size = colSums(reads_single) * norm_factors_single)
  #### output the phase specific scores (mean of normalized expression levels in the phase)
  apply(reads_single_cpm,2,mean)

})

#### normalization function
flexible_normalization <- function(data_in,by_row=TRUE){
  if(by_row){
    row_mean <- apply(data_in,1,mean)
    row_sd   <- apply(data_in,1,sd)
    output <- data_in
    for(i in 1:dim(data_in)[1]){
      output[i,] <- (data_in[i,] - row_mean[i])/row_sd[i]
    }
  }
  #### if by column
  if(!by_row){
    col_mean <- apply(data_in,2,mean)
    col_sd   <- apply(data_in,2,sd)
    output <- data_in
    for(i in 1:dim(data_in)[2]){
      output[,i] <- (data_in[,i] - col_mean[i])/col_sd[i]
    }
  }
  output
}

#### apply the normalization function
## first normalized for each phase
ans_normed <- flexible_normalization(ans,by_row=FALSE)
## then normalized of each cell
ans_normed_normed <- flexible_normalization(ans_normed,by_row=TRUE)

cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])
assign_cell_phase <- data.frame(cell_phase)
```


The cell phase vector is given by 

```{r echo=TRUE, eval=TRUE}
cell_phase_vector <- as.vector(as.matrix(assign_cell_phase));
```


## Maximum normalized value of cell cycle scores

### Admixture model (topic model)

We now fit an admixture model (topic model) on the molecule count data comprising 
of only the cell cycle genes.

```{r sructure, echo=TRUE, eval=TRUE, cache=TRUE}
structure_cell_cycle_genes <- function(data,K, type_info)
{
    Topic_clus <- topics(data, K = K, tol = 0.005)
    docweights <- Topic_clus$omega;
    type_ordered <- type_info[order(type_info)];
    docweights_ordered <- docweights[order(type_info),];
    barplot(t(docweights_ordered),col=2:(K+1),axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=0.3,cex.main=1.4);
    labels = match(unique(type_ordered), type_ordered);
    abline(v=labels-1)
    
    labels_low=labels-1;
    labels_up=c(labels_low[2:length(labels_low)],dim(docweights_ordered)[1]);
    mid_point <- labels_low +0.5*(labels_up-labels_low);
    axis(1,at=mid_point, unique(type_ordered),las=2,cex.axis=0.3);
}

sapply(2:7, function(k) structure_cell_cycle_genes(t(molecules_single_cell_cycle),
                                                   k,cell_phase_vector))
```


### t-SNE

We now implement t-SNE on the molecule counts based on the data on the cell cycle genes to see how the single cells cluster

```{r tsne, echo=TRUE, eval=TRUE, cache=FALSE}
tsne_samples <- tsne(t(molecules_single_cell_cycle),2);
iplot(tsne_samples[,1],tsne_samples[,2],rep(1,length(tsne_samples[,1])),cell_phase_vector)
```


## Maximum absolute normalized value of cell cycle scores

From the admixture plot on the cell cycle genes, there seems to be a weak cell cycle specific effect indeed even for the cell cycle genes. We now focus on a different assignment scheme where instead of assigning the cell to the maximum normalized value, we assign it to the maximum absolute normalized value. 

### Admixture model

```{r echo=TRUE, eval=TRUE, cahce=TRUE}
cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(abs(x))])
assign_cell_phase <- data.frame(cell_phase)

cell_phase_vector <- as.vector(as.matrix(assign_cell_phase));

sapply(2:7, function(k) structure_cell_cycle_genes(t(molecules_single_cell_cycle),k,cell_phase_vector))
```

### t-SNE plot

The t-sne plot with modified cell phase info is the following

```{r echo=TRUE, eval=TRUE}
iplot(tsne_samples[,1],tsne_samples[,2],rep(1,length(tsne_samples[,1])),cell_phase_vector)
```

