---
title: "Exploratory analysis: ECDF"
author: "Joyce Hsiao"
date: "2016-01-28"
output: 
  html_document:
    css: floating-toc.css
    toc: true
---


## Background and objectives

Explore visually the empirical distribution function of count in each individual. Perhaps this can give me some perspective on how to compare distributions in single-cell RNA-seq data.


## Some thoughts

Counting process may be useful for modeling single-cell sequencing data. In fact, the ECDF lines remind me of survival curves... 


## Setting up

```{r}
library(Humanzee)
```

Import filtered molecule counts

```{r}
molecules_filter <- read.table("../data/gilad-2015/molecules-filter.txt",
                               header = TRUE,
                               stringsAsFactors = FALSE)
dim(molecules_filter)
```

Import annotations of the filtered data

```{r}
anno_filter <- read.table("../data/gilad-2015/annotation-filter.txt",
                               header = TRUE,
                               stringsAsFactors = FALSE)
dim(anno_filter)
```


## Cumulative distribution


```{r}
load(file = "../data/gilad-2015/adj-cv.rda")
gene_order_mad <- order(mad, decreasing = TRUE)
gene_counts_order_mad <- molecules_filter[gene_order_mad, ]

anno_filter_r3 <- anno_filter[ which(anno_filter$replicate == "r3"), ]
gene_counts_order_mad_r3 <- gene_counts_order_mad[ , anno_filter_r3$replicate == "r3"]
                                                   
par(mfrow = c(1,2))
cols <- c("red", "green", "blue")
individuals <- unique(anno_filter$individual)
plot( ecdf(
    gene_counts_order_mad_r3[200, 
                          anno_filter_r3$individual == individuals[1] ] ),
    main = "", col = cols[1], xlim = c(20, 130))
for (ii_individual in c(2:length(individuals))) {
lines( ecdf( gene_counts_order_mad_r3[200, 
          anno_filter_r3$individual == individuals[ii_individual] ]), 
          col = cols[ii_individual] ) 
}
density_overlay(gene_counts_order_mad_r3,
             annotation = anno_filter_r3,
             which_gene = rownames(gene_counts_order_mad_r3)[200],
             xlab = "Count")
```




## Session information

```{r}
sessionInfo()
```

