---
title: 'LCL analysis'
author: "Kushal K Dey & Joyce Hsiao"
date: "2015-09-23"
output: 
  html_document:
    toc: true
---

## Objective

In this script, we perform Structure on LCL data and try to cluster the LCLs into sub types taking into account the cell phase data as well as the batch or sequencing lane information. 


## Set up

```{r, message=FALSE, warning=FALSE}
library(readr)
library(data.table)
library(devtools)
library(CountClust)
library(parallel)
library(dplyr)
library(ggplot2)
library(edgeR)
library(gplots)
```

## Prepare data

```{r}
lcl_qc_data <- read_csv("../data/qc-lcl.csv")
dim(lcl_qc_data)

lcl_annotations <- read.table('../data/annotation-lcl.txt',header=TRUE)
dim(lcl_annotations)
```

Read LCL data of both single and bulk data.

```{r echo=TRUE, eval=TRUE}
lcl_molecules <- t(data.frame(fread('../data/molecules-lcl.txt'), row.names = 1))
dim(lcl_molecules)
```

Extract the 96 LCLs by matching the lane name with the batch ids.

```{r}
lcl_indices <- match(lcl_qc_data$ll_name, lcl_annotations$well)
lcl_molecules_refined <- lcl_molecules[lcl_indices,]
lcl_annotations_refined <- lcl_annotations[lcl_indices,]
head(cbind(as.character(lcl_annotations_refined$well), rownames(lcl_molecules_refined)), 13)
```

Remove wells with no cells in them.

```{r}
ii_presentWell <- which(lcl_qc_data$cell.num > 0)

lcl_molecules_refined <- lcl_molecules_refined[ii_presentWell, ]
lcl_annotations_refined <- lcl_annotations_refined[ii_presentWell,]
stopifnot(nrow(lcl_molecules_refined) == nrow(lcl_annotations_refined))

dim(lcl_molecules_refined)
```

Remove ERCC genes

```{r}
ii_nonERCC <- grep("ERCC", colnames(lcl_molecules_refined), invert = TRUE)

lcl_molecules_refined <- lcl_molecules_refined[ , ii_nonERCC]
```

```{r}
range(lcl_molecules_refined)
```

Identify overexpressed genes.

```{r}
ii_notOverexpressed <- apply(lcl_molecules_refined, 2, function(x) any(x < 1024) )
lcl_molecules_refined <- lcl_molecules_refined[ , ii_notOverexpressed]
dim(lcl_molecules_refined)
```

Creating the capture site information from the well ids 

```{r}
well_id <- lcl_annotations_refined$well
ids <- sapply(well_id, function(x) substring(x,2,3))
batch_lab <- array(0, length(ids))
batch_lab[which(ids=="01" | ids=="02" | ids=="03")] <- 1
batch_lab[which(ids=="04" | ids=="05" | ids=="06")] <- 2
batch_lab[which(ids=="07" | ids=="08" | ids=="09")] <- 3
batch_lab[which(ids=="10" | ids=="11" | ids=="12")] <- 4
batch_lab <- factor(batch_lab, levels = c(1:4))
```


## Admixture modelling

We install the package CountClust and then apply the Structure analysis 
on the LCL cells taking into account the sample metadata owing to individual id, 
the batch id and the number of cells in the well as well. 

```{r}
library(CountClust)
library(maptpx)

# Remove rows with more than 1 missing values
lcl_molecules_refined <- handleNA(lcl_molecules_refined)$data


# Try fitting 2 to 5 clusters 
nclust_vec <- 2:5
if(file.exists("../rdas/lcl-structure.rda")) {
    load("../rdas/lcl-structure.rda")
} else {
    topics_list <- lapply(nclust_vec, function(per_clust) {
        topics( lcl_molecules_refined, K = per_clust, tol = 0.005)
        })    
    names(topics_list) <- paste0("clust", nclust_vec)
    save(topics_list, 
         file = "../rdas/lcl-structure.rda")
}
```



Structure plots

```{r}
nclust <- length(topics_list)
type_info <- batch_lab

for (per_clust in 1:nclust) {
    topics_clust <- topics_list[[per_clust]]
    clust <- topics_clust$K
    docweights <- topics_clust$omega
    type_ordered <- type_info[order(type_info)]
    docweights_ordered <- docweights[order(type_info),]
    barplot( t(docweights_ordered), 
             col = 2:(clust+1), axisnames = F,
             space = 0, border=NA,
             main = paste("No. of clusters=", clust), las = 1,
             ylim = c(0,1), cex.axis = 0.3, cex.main=1.4)
    labels <- match(unique(type_ordered), type_ordered)
    abline(v=labels-1)
    labels_low <- labels-1
    labels_up <- c(labels_low[2:length(labels_low)], dim(docweights_ordered)[1])
    mid_point <- labels_low +0.5*(labels_up-labels_low)
    axis(1, at = mid_point, unique(type_ordered), las = 2, cex.axis=0.3)
}
```

It seems from the above plots there is batch effect in the data and that is most clear for number of topics equal to 4 or 5. It also seems all the batches have very distinct patterns. Now we obtain batch corrected counts.


### Batch Corrected Model (single + bulk)

```{r batch_correct, echo=TRUE, eval=TRUE}

batch_corrected_lcl_molecules <- BatchCorrectedCounts(lcl_molecules_refined,batch_lab,use_parallel=TRUE)
```

Now we apply StructureObj on the batch corrected new counts data and see what patterns we get from the new data analysis, ideally we should not see the batch patterns that we saw in the previous run.

```{r batch_corrected_structure, echo=TRUE, eval=TRUE}

nclus_vec <- 2:5;
samp_metadata <- cbind.data.frame(1:96);
colnames(samp_metadata) = c("samples");
batch_lab <- batch_lab

if(!dir.exists("../../figures/lcl_structure_batchcorrect_bulk")) dir.create("../../figures/lcl_structure_batchcorrect_bulk")

#StructureObj(batch_corrected_lcl_molecules,nclus_vec,samp_metadata = samp_metadata, tol=0.001, batch_lab = NULL, path_rda="../../../project/rdas/topic_fit_lcl_batchcorrect_bulk.rda",partition=c('TRUE'),path_struct = "../../figures/lcl_structure_batchcorrect_bulk");


lcl_topics_batchcorrect <- get(load("../../../project/rdas/topic_fit_lcl_batchcorrect_bulk.rda"));


for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(lcl_topics_batchcorrect[[num]]$omega, samp_metadata = samp_metadata, batch_lab = batch_lab,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct='../../figures/lcl_structure_batchcorrect_bulk',control=list(cex.axis=1));
}

```


The batch removed Structure plots are as follows 

<img src='../../figures/lcl_structure_batchcorrect_bulk/clus_2/struct_clus_2_batch.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_bulk/clus_3/struct_clus_3_batch.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_bulk/clus_4/struct_clus_4_batch.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_bulk/clus_5/struct_clus_5_batch.png' style="width:304px;height:228px;">

We now assign cell phase to each of the 96 cells and try to see if any cell phase effects are seen in the Structure plot analysis or not. 

We first extract the LCL single cells. 


### Extracting LCL single cells 

```{r lcl_single_cells, echo=TRUE, eval=TRUE}

lcl_single_cells <- lcl_molecules_refined[which(lcl_qc_data$cell.num==1),];
```

We find there are `{r dim(lcl_single_cells)[1]} LCL single cells and we also extract the batch labels for these LCLs.

```{r batch_single_cells, echo=TRUE, eval=TRUE}

batch_lab_single_cells <- batch_lab[which(lcl_qc_data$cell.num==1)];

```


Now we assign each of these cells to a particular cell phase. For this we use the LCL reads level data and the cell cycle gene list from the previous single scell study to obtain the cell cycle phases for the single cells. 

```{r cell_cycle_genes, echo=TRUE, eval=TRUE}

cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})


```

We next assign the cell phase scores for each phase to the single cell data and then normalize the results. 

```{r echo=TRUE, eval=TRUE, results='hide'}

reads_lcl <- data.frame(fread('../data/reads-lcl.txt'),row.names=TRUE);
reads_lcl <- reads_lcl[,lcl_indices];
reads_lcl_single <- reads_lcl[,which(lcl_qc_data$cell.num==1)];
reads_single <- as.matrix(reads_lcl_single)


ans <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
  #### add average expression of all genes in the phase
  combined_matrix <- rbind(reads_single_phase,average=apply(reads_single_phase,2,mean))
  #### use transpose to compute cor matrix
  suppressWarnings(cor_matrix <- cor(t(combined_matrix)))
  #### take the numbers
  cor_vector <- cor_matrix[,dim(cor_matrix)[1]]
  #### restrict to correlation >= 0.3 
  reads_single_phase_restricted <- reads_single_phase[rownames(reads_single_phase) %in% names(cor_vector[cor_vector >= 0.3]),]
  #### apply normalization to reads
  norm_factors_single <- calcNormFactors(reads_single_phase_restricted, method = "TMM")
  reads_single_cpm <- cpm(reads_single_phase_restricted, log = TRUE,
                            lib.size = colSums(reads_single) * norm_factors_single)
  #### output the phase specific scores (mean of normalized expression levels in the phase)
  apply(reads_single_cpm,2,mean)

})

flexible_normalization <- function(data_in,by_row=TRUE){
  if(by_row){
    row_mean <- apply(data_in,1,mean)
    row_sd   <- apply(data_in,1,sd)
    output <- data_in
    for(i in 1:dim(data_in)[1]){
      output[i,] <- (data_in[i,] - row_mean[i])/row_sd[i]
    }
  }
  #### if by column
  if(!by_row){
    col_mean <- apply(data_in,2,mean)
    col_sd   <- apply(data_in,2,sd)
    output <- data_in
    for(i in 1:dim(data_in)[2]){
      output[,i] <- (data_in[,i] - col_mean[i])/col_sd[i]
    }
  }
  output
}

#### apply the normalization function
## first normalized for each phase
ans_normed <- flexible_normalization(ans,by_row=FALSE)
## then normalized of each cell
ans_normed_normed <- flexible_normalization(ans_normed,by_row=TRUE)

cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])
assign_cell_phase <- data.frame(cell_phase)

```

We now use the cell-phase to rearrange the Structure plots both before and after the batch correction. First we do the before batch correction (organized by cell phase)

```{r, echo=TRUE, eval=TRUE}

samp_metadata <- cbind.data.frame(as.vector(as.matrix(assign_cell_phase)));
colnames(samp_metadata) = c("cell_phase");

nclus_vec <- 2:5;

if(!dir.exists("../../figures/lcl_structure_single")) dir.create("../../figures/lcl_structure_single")

#StructureObj(lcl_single_cells ,nclus_vec,samp_metadata = samp_metadata, tol=0.001, batch_lab = NULL, path_rda="../../../project/rdas/topic_fit_lcl_single.rda",partition=c('TRUE'),path_struct = "../../figures/lcl_structure_single");


lcl_topics_single <- get(load("../../../project/rdas/topic_fit_lcl_single.rda"));


for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(lcl_topics_single[[num]]$omega, samp_metadata = samp_metadata, batch_lab = NULL,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct='../../figures/lcl_structure_single',control=list(cex.axis=1));
}
```

<img src='../../figures/lcl_structure_single/clus_2/struct_clus_2_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_single/clus_3/struct_clus_3_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_single/clus_4/struct_clus_4_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_single/clus_5/struct_clus_5_cell_phase.png' style="width:304px;height:228px;">

Now we perform the batch corrected counts.

```{r echo=TRUE, eval=TRUE}

batch_corrected_lcl_single <- BatchCorrectedCounts(lcl_single_cells,batch_lab_single_cells,use_parallel=TRUE)

```

Now we use the Structure model on the batch corrected counts.

```{r echo=TRUE, eval=TRUE}

nclus_vec <- 2:5;
samp_metadata <- cbind.data.frame(as.vector(as.matrix(assign_cell_phase)));
colnames(samp_metadata) = c("cell_phase");

if(!dir.exists("../../figures/lcl_structure_batchcorrect_single")) dir.create("../../figures/lcl_structure_batchcorrect_single")

#StructureObj(batch_corrected_lcl_single,nclus_vec,samp_metadata = samp_metadata, tol=0.001, batch_lab = NULL, path_rda="../../../project/rdas/topic_fit_lcl_batchcorrect_single.rda",partition=c('TRUE'),path_struct = "../../figures/lcl_structure_batchcorrect_single");


lcl_topics_batchcorrect_single <- get(load("../../../project/rdas/topic_fit_lcl_batchcorrect_single.rda"));


for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(lcl_topics_batchcorrect_single[[num]]$omega, samp_metadata = samp_metadata, batch_lab = NULL,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct='../../figures/lcl_structure_batchcorrect_single',control=list(cex.axis=1));
}

```

<img src='../../figures/lcl_structure_batchcorrect_single/clus_2/struct_clus_2_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_single/clus_3/struct_clus_3_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_single/clus_4/struct_clus_4_cell_phase.png' style="width:304px;height:228px;">
<img src='../../figures/lcl_structure_batchcorrect_single/clus_5/struct_clus_5_cell_phase.png' style="width:304px;height:228px;">

```{r}
sessionInfo()
```

