---
title: "classtpx on Leng data using Buettner data"
author: "Kushal K Dey"
date: "March 24, 2016"
output: html_document
---

## Introduction

In this script, we aim to perform `classtpx` on the data due to [Leng et al 2015](http://www.nature.com/nmeth/journal/v12/n10/full/nmeth.3549.html) based on the data due to 
[Buettner et al 2015](http://www.nature.com/nbt/journal/v33/n2/full/nbt.3102.html). The datasets for these two datasets are saved as expressionSet objects on our Github repositories [Buettner data](https://github.com/jhsiao999/singleCellRNASeqMouseBuettnerESC) and [Leng data](https://github.com/jhsiao999/singleCellRNASeqHumanLengESC).

## Prepare Leng 2015 data

```{r echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
# devtools::install_github("kkdey/singleCellRNASeqHumanLengESC", force=TRUE)
library(singleCellRNASeqHumanLengESC)
data("HumanLengESC")
leng_gene_names <- Biobase::featureNames(HumanLengESC);

leng_data <- Biobase::exprs(HumanLengESC);
leng_metadata <- Biobase::pData(HumanLengESC)
leng_cell_state <- leng_metadata$cell_state;
```


## Prepare the Buettner 2015 data

```{r echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
# devtools::install_github("jhsiao999/singleCellRNASeqMouseBuettnerESC", force=TRUE)
library(singleCellRNASeqMouseBuettnerESC)
data("MouseBuettnerESC")
buettner_gene_names <- Biobase::featureNames(MouseBuettnerESC);

buettner_data <- Biobase::exprs(MouseBuettnerESC);
buettner_metadata <- Biobase::pData(MouseBuettnerESC)
buettner_cell_state <- buettner_metadata$cell_cycle;
```

## Ortholog Human genes for Buettner mouse genes 

```{r echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
library(biomaRt)
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl");
attributes = c("ensembl_gene_id","mmusculus_homolog_ensembl_gene","mmusculus_homolog_perc_id_r1")
attributes=c(attributes,"mmusculus_homolog_orthology_type",
"mmusculus_homolog_subtype", "mmusculus_homolog_perc_id")
orth.mouse.human = getBM(attributes,
filters="with_homolog_mmus",values=TRUE, mart = human,
uniqueRows=TRUE)

```

Match the gene names observed here with the list from BioMart.

```{r echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

filter_indices <- match(buettner_gene_names, orth.mouse.human[,2]);
filter_indices <- filter_indices[!is.na(filter_indices)];
ortholog_buettner_mouse_ids <- orth.mouse.human[filter_indices,2];
ortholog_buettner_human_ids <- orth.mouse.human[filter_indices,1];

human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mySymbols <- ortholog_buettner_human_ids; # mySymbols is a vector of MGI symbols.
gene_list_human <- getBM( attributes=c("ensembl_gene_id", "hgnc_symbol") , filters=
"ensembl_gene_id", values =mySymbols ,mart=human)

gene_names_human <- gene_list_human[,2];

```

```{r echo=TRUE, eval=TRUE}

matched_human_gene_indices_leng <- match(gene_names_human, leng_gene_names)
matched_human_gene_indices_leng <- matched_human_gene_indices_leng[!is.na(matched_human_gene_indices_leng)];


matched_genes_leng <- leng_gene_names[matched_human_gene_indices_leng];

filter_indices <- match(matched_genes_leng, leng_gene_names);

leng_filtered_data <- leng_data[filter_indices,];
```

Match the profile with the Buettner data

```{r echo=TRUE, eval=TRUE}

filter_indices <- match(gene_names_human, leng_gene_names);
gene_list_human_filtered <- gene_list_human[which(!is.na(filter_indices)),];

gene_names_human_filtered <- gene_list_human_filtered[,1];
human_ids_buettner <- ortholog_buettner_human_ids[match(gene_names_human_filtered, ortholog_buettner_human_ids)];

mouse_ortholog_ids_buettner <- ortholog_buettner_mouse_ids[match(gene_names_human_filtered, ortholog_buettner_human_ids)];

filter_genes_buettner <- match(mouse_ortholog_ids_buettner, buettner_gene_names);

buettner_filtered_data <- buettner_data[filter_genes_buettner,];

```

We pool the data for the Buettner data and the Leng data.

```{r echo=TRUE, eval=TRUE}
pooled_data <- t(cbind(buettner_filtered_data, leng_filtered_data));
```

## maptpx application

```{r echo=TRUE, eval=FALSE}
Topic_clus <- maptpx::topics(
    pooled_data, 
    K=3, 
    tol=0.1
    )

save(Topic_clus, file="../rdas/leng_topic_fit_3_maptpx_buettner.rda")

```

```{r echo=TRUE, eval=TRUE, warning=FALSE, fig.show="asis", dpi=144, fig.width=3, fig.height=5}
Topic_clus <- get(load(file="../rdas/leng_topic_fit_3_maptpx_buettner.rda"))

omega <- Topic_clus$omega;

leng_cell_state_human <- paste0("human", "_", leng_cell_state);
buettner_cell_state_mouse <- paste0("mouse","_", buettner_cell_state);

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = factor(c(as.character(buettner_cell_state_mouse), as.character(leng_cell_state_human)),
                        levels = c("mouse_G1", "mouse_S", "mouse_G2M",
                                        "human_G1","human_S",
                                         "human_G2", "human_H1") ) )


rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))
```

## classtpx application

```{r echo=TRUE, eval=TRUE}
known_samples <- 1:288;
class_labs <- c(rep(1,96), rep(2,96), rep(3,96));
```

### omega.fix method 

```{r echo=TRUE, eval=FALSE}

Topic_clus <- classtpx::class_topics(
    pooled_data, 
    K=3, 
    known_samples = known_samples,
    class_labs = class_labs,
    method="omega.fix",
    tol=0.01)

save(Topic_clus, file="../rdas/leng_topic_fit_classtpx_omega_fix_buettner.rda")
```

```{r echo=TRUE, eval=TRUE, warning=FALSE, fig.show="asis", dpi=144, fig.width=3, fig.height=5}
Topic_clus <- get(load(file="../rdas/leng_topic_fit_classtpx_omega_fix_buettner.rda"))

omega <- Topic_clus$omega;

leng_cell_state_human <- paste0("human", "_", leng_cell_state);
buettner_cell_state_mouse <- paste0("mouse","_", buettner_cell_state);

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = factor(c(as.character(buettner_cell_state_mouse), as.character(leng_cell_state_human)),
                        levels = c("mouse_G1", "mouse_S", "mouse_G2M",
                                        "human_G1","human_S",
                                         "human_G2", "human_H1") ) )


rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))
```

### theta.fix method 

```{r echo=TRUE, eval=FALSE}

Topic_clus <- classtpx::class_topics(
    pooled_data, 
    K=3, 
    known_samples = known_samples,
    class_labs = class_labs,
    method="theta.fix",
    tol=0.01,
    shrink=FALSE)

save(Topic_clus, file="../rdas/leng_topic_fit_classtpx_theta_fix_buettner.rda")
```

```{r echo=TRUE, eval=TRUE, warning=FALSE, fig.show="asis", dpi=144, fig.width=3, fig.height=5}
Topic_clus <- get(load(file="../rdas/leng_topic_fit_classtpx_theta_fix_buettner.rda"))

omega <- Topic_clus$omega;

leng_cell_state_human <- paste0("human", "_", leng_cell_state);
buettner_cell_state_mouse <- paste0("mouse","_", buettner_cell_state);

annotation <- data.frame(
  sample_id = paste0("X", c(1:NROW(omega))),
  tissue_label = factor(c(as.character(buettner_cell_state_mouse), as.character(leng_cell_state_human)),
                        levels = c("mouse_G1", "mouse_S", "mouse_G2M",
                                        "human_G1","human_S",
                                         "human_G2", "human_H1") ) )


rownames(omega) <- annotation$sample_id;


CountClust::StructureGGplot(omega = omega,
                annotation = annotation,
                palette = RColorBrewer::brewer.pal(8, "Accent"),
                yaxis_label = "Cell cycle phase",
                order_sample = TRUE,
                axis_tick = list(axis_ticks_length = .1,
                                 axis_ticks_lwd_y = .1,
                                 axis_ticks_lwd_x = .1,
                                 axis_label_size = 7,
                                 axis_label_face = "bold"))
```
