---
title: "Store some published data in ExpressionSet objects"
author: "Joyce Hsiao"
date: "2016-02-23"
output: 
  html_document:
    css: floating-toc.css
    toc: true
---


```{r, include = FALSE}
library(knitr)
opts_chunk$set(eval = FALSE, echo = TRUE)
```


## Background and objectives

Make ExpressionSet objects for some publishe datasets.


## Load packages

```{r}
library(data.table)
library(Biobase)
```




## Battles et al., 2014, Human QTL, multiple data types


```{r}
##<----- Ribosome profiling assay ----->##

# import count matrix, convert to a matrix object
counts_ribo <- 
    read.table('../data/QTLHumanBattles/GSE61742_YRI_RPF_count_by_gene.table.txt',
               sep = "\t", header = TRUE,
               stringsAsFactors = FALSE,
               row.names = 1)
counts_ribo[1:5, 1:5]
dim(counts_ribo)

# feature info
row_data <- data.frame(gene_name = rownames(counts_ribo))
rownames(row_data) <- rownames(counts_ribo)
    
# phenotype info
meta_list <- data.frame(
    labelDescription = c("Haptype Yoruba individuals"),
    row.names = c("individual"))
col_data <- data.frame(
    individual = gsub("[[:alpha:]]", "", colnames(counts_ribo)),
    stringsAsFactors = FALSE)
rownames(col_data) <- col_data$individual
colnames(counts_ribo) <- col_data$individual

stopifnot(all.equal(rownames(col_data), colnames(counts_ribo)))

# putting all together
HumanBattlesRibo <- new("ExpressionSet",
    exprs = as.matrix(counts_ribo),
    phenoData = new("AnnotatedDataFrame",
                    data = col_data,
                    varMetadata = meta_list),
    featureData = new("AnnotatedDataFrame",
                      data = row_data),
    experimentData = new("MIAME",
                         title = "Ribosome profiling data of 72 Yoruba indivdiuals"))


##<----- RNA-seq data ----->##
# http://eqtl.uchicago.edu/RNA_Seq_data/results/

# import count matrix, convert to a matrix object
counts_rna <- 
    read.table('../data/QTLHumanBattles/final_gene_counts',
               sep = " ", header = TRUE,
               stringsAsFactors = FALSE,
               row.names = 1)
counts_rna[1:5, 1:5]
dim(counts_rna)

# remove a column of NAs
counts_rna <- counts_rna[ , -dim(counts_rna)[2]]

# feature info
row_data <- data.frame(gene_name = rownames(counts_rna),
                       chr = counts_rna[ ,1],
                       len = counts_rna[ ,2])
rownames(row_data) <- rownames(counts_rna)

# phenotype info
meta_list <- data.frame(
    labelDescription = c("Haptype Yoruba individuals",
                         "Replicate",
                         "Sequencing center",
                         "Unique sample ID"),
    row.names = c("individual",
                  "replicate",
                  "seq_site",
                  "sample_id") )
col_strings <- colnames(counts_rna)[-c(1:2)]
col_strings <- gsub("NA", "", col_strings)



col_data <- data.frame(
    individual = substr(col_strings, start = 1, stop = 5),
    replicate = unlist(sapply(strsplit(col_strings, split = "_"),
                       function(x) {
                          if (length(x) == 2) return(1)
                          if (length(x) == 3) return(2)
                      }) ), 
    seq_site = unlist( sapply(strsplit(col_strings, split = "_"),
                       function(x) x[length(x)] ) ),
    sample_id = colnames(counts_rna)[-c(1:2)],
    stringsAsFactors = FALSE)
rownames(col_data) <- colnames(counts_rna)[-c(1:2)]


# putting all together
HumanBattlesRNA <- new("ExpressionSet",
    exprs = as.matrix(counts_rna)[ , -c(1:2)],
    phenoData = new("AnnotatedDataFrame",
                    data = col_data,
                    varMetadata = meta_list),
    featureData = new("AnnotatedDataFrame",
                      data = row_data),
    experimentData = new("MIAME",
                         title = "RNA-seq data of 72 Yoruba indivdiuals"))


##<----- QTLs ----->##
library(gdata)
eQTL <- read.xls("../data/1260793_DatafileS1.xlsx",
                 sheet = 2)

rQTL <- read.xls("../data/1260793_DatafileS1.xlsx",
                 sheet = 3)

pQTL <- read.xls("../data/1260793_DatafileS1.xlsx",
                 sheet = 4)
dim(eQTL)
dim(rQTL)
dim(pQTL)


# phenotype info
meta_list <- data.frame(
    labelDescription = c("empirical gene-level p-value for the most
                          significant SNP basd on 10,000 permutations of
                          genotype sample labels",
                         "Pearson correlation p-value between the exprssion
                          phenotype and genotype at the most significant SNP
                          for the gene",
                         "Pearson correlation between the expression phenotype
                         and the genotype at the most significant SNP for the gene",
                         "position of the most significant SNP"),
    row.names = c("perm.p.value",
                  "snp.pvalue",
                  "snp.R.value",
                  "hg18.pos") )



# putting all together
QTLHumanBattlesRNA <- new("ExpressionSet",
    exprs = as.matrix(eQTL[ , -c(1:2)]),
    featureData = new("AnnotatedDataFrame",
                      data = data.frame(gene_names = eQTL[,1],
                                        chr = eQTL[,2],
                                        row.names = eQTL[,1],
                                        stringsAsFactors = FALSE)),
    experimentData = new("MIAME",
                         title = "eQTL"))


QTLHumanBattlesRibo <- new("ExpressionSet",
    exprs = as.matrix(rQTL[ , -c(1:2)]),
    featureData = new("AnnotatedDataFrame",
                      data = data.frame(gene_names = rQTL[,1],
                                        chr = rQTL[,2],
                                        row.names = rQTL[,1],
                                        stringsAsFactors = FALSE)),
    experimentData = new("MIAME",
                         title = "ribosome-QTL"))


QTLHumanBattlesProtein <- new("ExpressionSet",
    exprs = as.matrix(pQTL[ , -c(1:2)]),
    featureData = new("AnnotatedDataFrame",
                      data = data.frame(gene_names = pQTL[,1],
                                        chr = pQTL[,2],
                                        row.names = pQTL[,1],
                                        stringsAsFactors = FALSE)),
    experimentData = new("MIAME",
                         title = "protein-QTL"))


# combine three datasets
# gene_names <- as.character(intersect(
#                            intersect(eQTL[,1], rQTL[,1]), pQTL[,1]))
# length(gene_names)
# gene_names <- as.character(intersect(eQTL[,1], rQTL[,1]))
# length(gene_names)
# eQTL + rQTL
# merge(eQTL[ , c("ENSG", "snp.pvalue")],
#               rQTL[ , c("ENSG", "snp.pvalue")], by = "ENSG")

save(HumanBattlesRibo, HumanBattlesRNA,
     QTLHumanBattlesRNA, QTLHumanBattlesRibo, QTLHumanBattlesProtein,
        file = "../data/all-rda/QTLHumanBattles.rda")
```



## Session information
```{r}
sessionInfo()
```

