---
title: 'iPSC clustering: are there hidden cell cycle effects?'
author: "Kushal K Dey, Joyce H. Siao"
date: "July 21, 2015"
output: html_document
---

When we performed Structure on the single cell - gene data table of counts, with the aim to cluster the single cells, we observed that the individual effets largely drove the clusters and the cell cycle specific effects did not show up. A general feeling is that if we account for the individual effect as a batch effect and redo our analysis, we should get the cell cycle information getting reflected in the cluster patterns. But is it true?...What if we do the analysis and then find that the cell cycle specific effects do not show up?...Can we be little more confident about what whether we would get some cell cycle driven clusters once we use individual information as a batch effect in our statistical model. One way to figure this out would be to cluster the single cells by the Structure model 

```{r, echo=FALSE, eval=TRUE}
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(maptpx)))
suppressMessages(suppressWarnings(library(gplots)))
suppressMessages(suppressWarnings(library(philentropy)))
suppressMessages(suppressWarnings(library(dplyr)))
suppressMessages(suppressWarnings(library(edgeR)))


#setwd('/Users/kushal/Documents/Matthew Stephens Project/counts_clustering/Single_Cell_seq_Blischak/src/');
reads <- data.frame(fread('../data/reads.txt'),row.names=1);
molecules <-  data.frame(fread('../data/molecules.txt'),row.names=1);


quality_single_cells <- scan("../data/quality-single-cells.txt",
                             what = "character")

anno <- data.frame(fread('../data/annotation.txt'));

molecules <- molecules[, grepl("bulk", colnames(molecules)) |
                         colnames(molecules) %in% quality_single_cells]
anno <- anno[anno$well == "bulk" | anno$sample_id %in% quality_single_cells, ]
stopifnot(ncol(molecules) == nrow(anno),
          colnames(molecules) == anno$sample_id)

reads <- reads[, grepl("bulk", colnames(reads)) |
                         colnames(reads) %in% quality_single_cells]
stopifnot(ncol(reads) == nrow(anno),
          colnames(reads) == anno$sample_id)


expressed <- rowSums(molecules[, anno$well == "bulk"]) > 0 &
             rowSums(molecules[, anno$well != "bulk"]) > 0
molecules <- molecules[expressed, ]
#dim(molecules)

expressed <- rowSums(reads[, anno$well == "bulk"]) > 0 &
             rowSums(reads[, anno$well != "bulk"]) > 0
reads <- reads[expressed, ]
#dim(reads)

stopifnot(nrow(ercc) == 92)
quality_single_cells <- scan("../data/quality-single-cells.txt",
                             what = "character")

## just filtering cells that meet the QC check

molecules_single <- molecules[, colnames(molecules) %in% quality_single_cells]

molecules_single <- molecules %>% select(-contains("bulk"))
reads_single <- reads %>% select(-contains("bulk"))

## Remove genes with max molecule numer larger than 1024

molecules_single <- molecules_single[apply(molecules_single,1,max) < 1024,];

molecules_single <- molecules_single[-grep("ERCC",rownames(molecules_single)),]; 

reads_single <- reads_single[-grep("ERCC",rownames(reads_single)),];





cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})

ans <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
  #### add average expression of all genes in the phase
  combined_matrix <- rbind(reads_single_phase,average=apply(reads_single_phase,2,mean))
  #### use transpose to compute cor matrix
  cor_matrix <- cor(t(combined_matrix))
  #### take the numbers
  cor_vector <- cor_matrix[,dim(cor_matrix)[1]]
  #### restrict to correlation >= 0.3 
  reads_single_phase_restricted <- reads_single_phase[rownames(reads_single_phase) %in% names(cor_vector[cor_vector >= 0.3]),]
  #### apply normalization to reads
  norm_factors_single <- calcNormFactors(reads_single_phase_restricted, method = "TMM")
  reads_single_cpm <- cpm(reads_single_phase_restricted, log = TRUE,
                            lib.size = colSums(reads_single) * norm_factors_single)
  #### output the phase specific scores (mean of normalized expression levels in the phase)
  apply(reads_single_cpm,2,mean)

})

#head(ans)

#### normalization function
flexible_normalization <- function(data_in,by_row=TRUE){
  if(by_row){
    row_mean <- apply(data_in,1,mean)
    row_sd   <- apply(data_in,1,sd)
    output <- data_in
    for(i in 1:dim(data_in)[1]){
      output[i,] <- (data_in[i,] - row_mean[i])/row_sd[i]
    }
  }
  #### if by column
  if(!by_row){
    col_mean <- apply(data_in,2,mean)
    col_sd   <- apply(data_in,2,sd)
    output <- data_in
    for(i in 1:dim(data_in)[2]){
      output[,i] <- (data_in[,i] - col_mean[i])/col_sd[i]
    }
  }
  output
}

#### apply the normalization function
## first normalized for each phase
ans_normed <- flexible_normalization(ans,by_row=FALSE)
## then normalized of each cell
ans_normed_normed <- flexible_normalization(ans_normed,by_row=TRUE)

#head(ans_normed_normed)

cell_phase_mat <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])
cell_phase <- as.vector(as.matrix(data.frame(cell_phase_mat)))

```

We first choose the single cells coming from the first individual.

```{r, echo=TRUE, eval=TRUE, cache=FALSE}

#cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])

structure_per_individual <- function(counts_data,individual_lab,nclus,cell_phase)
{
  subgroup_id =colnames(counts_data);
  individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
  batch = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));
 
 # indices <- which(individual_effect==unique(individual_effect)[individual_lab] & 
  #                   batch_effect==unique(batch_effect)[batch_lab] );
  
  indices <- which(individual_effect==unique(individual_effect)[individual_lab]);
  
  counts_ind <- counts_data[,indices];
  
  Topic_Clus <- topics(t(counts_ind), K=nclus, tol=0.005);
  
  cell_phase_ind <- cell_phase[indices];
  
  docweights_ind <- Topic_Clus$omega;
  
  cell_phase_ind_ordered <- cell_phase_ind[order(cell_phase_ind)];
  
  docweights_ind_ordered <- docweights_ind[order(cell_phase_ind),];
  
  barplot(t(docweights_ind_ordered),col=2:(nclus+1),axisnames=F,space=0,border=NA,main=paste("cellcycle ordering/ person=",individual_lab,"clusters=",nclus),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

  labels = match(unique(cell_phase_ind_ordered), cell_phase_ind_ordered);
  abline(v=labels-1)

  labels_low=labels;
  labels_up=c(labels[2:length(labels)],dim(docweights_ind_ordered)[1]);
  mid_point=labels_low +0.5*(labels_up-labels_low);

  axis(1,at=mid_point, unique(cell_phase_ind_ordered),las=2);
  
  batch_ind <- batch[indices];
  batch_ind_ordered <- batch_ind[order(batch_ind)];
  
  docweights_ind_ordered = docweights_ind[order(batch_ind),];

barplot(t(docweights_ind_ordered),col=2:(nclus+1),axisnames=F,space=0,border=NA,main=paste("Batch ordering/ person=",individual_lab,"clusters=",nclus),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(batch_ind_ordered), batch_ind_ordered);
abline(v=labels-1)

labels_low=labels-1;
labels_up=c(labels[2:length(labels)],dim(docweights_ind_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(batch_ind_ordered),las=2);

  

}


structure_per_individual(molecules_single,1,nclus=5,cell_phase)
structure_per_individual(molecules_single,2,nclus=5,cell_phase)
structure_per_individual(molecules_single,3,nclus=5,cell_phase)


```


This seenms like once we constrain for the individual effect, the batch information seems to be prominently showcased in the Structure plots.

What happens when we fix one individual and one batch?...Do we then get the cell cycle specific effects ??

We consider two different senarios with fixed individual and fixed batch. 

```{r echo=TRUE, eval=TRUE, cache=FALSE}
  counts_data <- molecules_single
  nclus <- 5
  subgroup_id =colnames(molecules_single);
  individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
  batch_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));
  individual_lab <- 1; batch_lab <- 1;
  indices <- which(individual_effect==unique(individual_effect)[individual_lab] & 
                     batch_effect==unique(batch_effect)[batch_lab] );
  
  counts_ind <- counts_data[,indices];
  
  Topic_Clus <- topics(t(counts_ind), K=nclus, tol=0.005);
  
  cell_phase_ind <- cell_phase[indices];
  
  docweights_ind <- Topic_Clus$omega;
  
  cell_phase_ind_ordered <- cell_phase_ind[order(cell_phase_ind)];
  
  docweights_ind_ordered <- docweights_ind[order(cell_phase_ind),];
  
  barplot(t(docweights_ind_ordered),col=2:(nclus+1),axisnames=F,space=0,border=NA,main=paste("cellcycle ordering/ person=",individual_lab,"batch=",batch_lab),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

  labels = match(unique(cell_phase_ind_ordered), cell_phase_ind_ordered);
  abline(v=labels-1)

  labels_low=labels;
  labels_up=c(labels[2:length(labels)],dim(docweights_ind_ordered)[1]);
  mid_point=labels_low +0.5*(labels_up-labels_low);

  axis(1,at=mid_point, unique(cell_phase_ind_ordered),las=2);
  

  

```


```{r echo=FALSE, eval=TRUE, cache=FALSE}
  counts_data <- molecules_single
  subgroup_id =colnames(counts_data);
  nclus <- 5
  individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
  batch_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));
  individual_lab <- 3; batch_lab <- 3;
  indices <- which(individual_effect==unique(individual_effect)[individual_lab] & 
                     batch_effect==unique(batch_effect)[batch_lab] );
  
  counts_ind <- counts_data[,indices];
  
  Topic_Clus <- topics(t(counts_ind), K=nclus, tol=0.005);
  
  cell_phase_ind <- cell_phase[indices];
  
  docweights_ind <- Topic_Clus$omega;
  
  cell_phase_ind_ordered <- cell_phase_ind[order(cell_phase_ind)];
  
  docweights_ind_ordered <- docweights_ind[order(cell_phase_ind),];
  
  barplot(t(docweights_ind_ordered),col=2:(nclus+1),axisnames=F,space=0,border=NA,main=paste("cellcycle ordering/ person=",individual_lab,"batch=",batch_lab),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

  labels = match(unique(cell_phase_ind_ordered), cell_phase_ind_ordered);
  abline(v=labels-1)

  labels_low=labels;
  labels_up=c(labels[2:length(labels)],dim(docweights_ind_ordered)[1]);
  mid_point=labels_low +0.5*(labels_up-labels_low);

  axis(1,at=mid_point, unique(cell_phase_ind_ordered),las=2);
  

  

```
