---
title: 'LCL data : cell ordering'
author: "Kushal K Dey"
date: "November 6, 2015"
output: 
  html_document:
    toc: true
---

```{r echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
library(qtlcharts)
library(CountClust)
library(parallel)
library(cellcycleR)
library(data.table)
library(binhf)
library(vioplot)
library(limma)
library(readxl)
library(readr)
library(mygene)
library(knitr)
```

## Objective 

In this script, we perform the cellcycleR method on the LCLs, which are not stem cells and may behave similar to the normal cells and show the growth phases, which we did not see in the stem cell studies we observed so far.


## Data Preparation

We first prepare the LCL single cell data. 

```{r echo=TRUE, eval=TRUE}
setwd('/Users/kushal/Documents/singleCell-method/project/analysis')
lcl_qc_data <- read_csv("../data/qc-lcl.csv");
lcl_annotations <- read.table('../data/annotation-lcl.txt',header=TRUE);
lcl_molecules <- t(data.frame(fread('../data/molecules-lcl.txt'), row.names = 1));

lcl_indices <- match(lcl_qc_data$ll_name, lcl_annotations$well)

lcl_molecules_refined <- lcl_molecules[lcl_indices,];
lcl_annotations_refined <- lcl_annotations[lcl_indices,];

well_id <- lcl_annotations_refined$well;
ids <- sapply(well_id, function(x) substring(x,2,3))
batch_lab <- array(0, length(ids))
batch_lab[which(ids=="01" | ids=="02" | ids=="03")]=1;
batch_lab[which(ids=="04" | ids=="05" | ids=="06")]=2;
batch_lab[which(ids=="07" | ids=="08" | ids=="09")]=3;
batch_lab[which(ids=="10" | ids=="11" | ids=="12")]=4;


lcl_single_cells <- lcl_molecules_refined[which(lcl_qc_data$cell.num==1),];
batch_lab_single_cells <- batch_lab[which(lcl_qc_data$cell.num==1)];
```

Now we focus only on the cell cycle genes listed in Macosko paper.

```{r echo=TRUE, eval=TRUE, results='hide'}
reads_lcl <- data.frame(fread('../data/reads-lcl.txt'),row.names=TRUE);
reads_lcl <- reads_lcl[,lcl_indices];
reads_lcl_single <- reads_lcl[,which(lcl_qc_data$cell.num==1)];
reads_single <- as.matrix(reads_lcl_single)

cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})

labs=unique(unlist(lapply(1:5, function(k) match(cell_cycle_genes_list[[k]],colnames(lcl_single_cells)))));
labs = labs[!is.na(labs)];
lcl_single_cell_cycle_genes <- lcl_single_cells[,labs];

batch_corrected_cell_cycle_genes <- BatchCorrectedCounts(lcl_single_cell_cycle_genes,batch_lab_single_cells,use_parallel=TRUE)

cycle_counts_data <-batch_corrected_cell_cycle_genes

dim(cycle_counts_data)

```

We now mean center and stanadardize the data for applying cellcycleR.

```{r echo=TRUE, eval=TRUE}
cycle_voom_data <- voom(cycle_counts_data)$E;
cycle_data_norm <- apply(cycle_voom_data,2,function(x)  return (x-mean(x))/sd(x))
celltime_levels <- 100;
cycle_data_norm <- cycle_data_norm[, -which(colSums(cycle_data_norm)==0)]

dim(cycle_data_norm)
```

## Applying cellcycleR on LCL data

```{r echo=TRUE, eval=FALSE}

out <- cell_ordering_class(cycle_data_norm, celltime_levels = 100, num_iter=100, save_path="../rdas/cell_order_lcl.rda")
```

We ran the method above once already (took around 2 minutes) and now we just load the output.

```{r echo=TRUE, eval=TRUE}
out <- get(load(file="../rdas/cell_order_lcl.rda"));
cell_order_full <- cell_ordering_full(out$signal_intensity, dim(cycle_data_norm)[2])
```

We look at the plots of the amplitudes, phases and the non signal variation of the genes.

```{r echo=TRUE, eval=TRUE}

amp_genes <- out$amp;
sd_genes <- out$sigma;
phi_genes <- out$phi;

plot(density(phi_genes), col="red", main="Density plot of the phases")
plot(density(amp_genes[-which.max(amp_genes)]), col="red", main="Density plot of the amplitudes")
plot(density(sd_genes[-which.max(sd_genes)]), col="red", main="Density plot of the non-signal sd")

```

We extract the genes with high SNR - these are more likely to be sinusoidal.

```{r echo=TRUE, eval=TRUE}
ESS <- amp_genes^2; RSS <- sd_genes^2

SNR <- ESS/RSS;

plot(SNR, col="red", pch=20, lwd=1)
top_genes <- which(SNR > 3);

```

Next we plot the qtlcharts for these top sinusoidal genes and see if their patterns are indeed
sinusoidal or not.

```{r echo=TRUE, eval=TRUE}
iplotCurves(t(cycle_data_norm[order(cell_order_full),top_genes]))
```

## Redoing on high SNR genes


We extract the high SNR genes as they seem to have sinusoidal patterns and repeat the procedure again. The aim here is to see how much robust the analysis is to the selction of the (most) sinusoidal genes.

```{r echo=TRUE, eval=TRUE}

snr_high_indices <- which(SNR > 1);

cycle_data_norm_sinusoidal <- cycle_data_norm[,snr_high_indices];

dim(cycle_data_norm_sinusoidal)
```

We apply the cell ordering mechanism (it takes around 5 minutes to run)

```{r echo=TRUE, eval=FALSE}
out2 <- cell_ordering_class(cycle_data_norm_sinusoidal, celltime_levels = 100, num_iter=100,
                            save_path="../rdas/cell_order_lcl_sinusoidal.rda")
```

We reload the output 

```{r echo=TRUE, eval=TRUE}
out2 <- get(load(file="../rdas/cell_order_lcl_sinusoidal.rda"));
cell_order_full <- cell_ordering_full(out2$signal_intensity, dim(cycle_data_norm_sinusoidal)[2])
```

We plot the same features as above and check for the robustness. We needed to shift the cell order so as to compare with previous plot on all genes as the method is non-identifiable upto a rotation. 

```{r echo=TRUE, eval=TRUE}

amp_genes <- out2$amp;
sd_genes <- out2$sigma;
phi_genes <- out2$phi;

plot(density(phi_genes), col="red", main="Density plot of the phases")
plot(density(amp_genes[-which.max(amp_genes)]), col="red", main="Density plot of the amplitudes")
plot(density(sd_genes[-which.max(sd_genes)]), col="red", main="Density plot of the non-signal sd")

ESS <- amp_genes^2; RSS <- sd_genes^2

SNR <- ESS/RSS;

plot(SNR, col="red", pch=20, lwd=1)
top_genes <- which(SNR > 3);
#new_cell_order <- shift(order(cell_order_full),1)
iplotCurves(t(cycle_data_norm_sinusoidal[order(cell_order_full),top_genes]))

```


## Observations

The LCL genes do not look as sinusoidal as the iPSC data looked. The sinusoidal assumption overall is not visible for many of the genes, despite them showing high SNR. We perform gene annotations to see if the method picks up genes that are meaningful and are related to cell cycle or not.

```{r echo=TRUE, eval=TRUE}
sinusoidal_gene_names <- colnames(cycle_data_norm_sinusoidal);
imp_genes <- sinusoidal_gene_names[order(SNR, decreasing=TRUE)]
top_genes_100 <- imp_genes[1:100];
out <- queryMany(top_genes_100,  scopes="ensembl.gene", fields=c("name","summary"), species="human");

kable(as.data.frame(out))

```

The gene annotations again seem to pick up quite a few cellcycle related genes. So, it does seem that the sinusoidal patterns in LCL are a bit weaker compared to iPSC and aftr oredering of the cells, not all genes look perfectly sinusoidal. But inspite of that, the method picks up genes associated with cell cycle and cell division, which is nice.
