---
title: "Poisson Batch Model on iPSC Cell cycle genes"
author: "Kushal K Dey & Joyce Hsiao"
date: "September 26, 2015"
output: html_document
---

## Goal

From the exploratory studies, it seemed that there are strong both batch and individual effects in the iPSC data that were resulting in subduing the cell cycle effects. So far, we have used BatchCorrectedCounts function from package CountClust to remove the batch effects in a preprocessing step and then transform the residuals to counts and apply the topic model on that count data. Here, we pool the individual and batch effect joint as a factor effect in a model based clusetring algorithm. Here we use the Poisson log link model and see whether this model is able to pick up the biological variation across the cell cycles or not.

```{r message = FALSE, warning = FALSE}
library(CountClust)
library(data.table)
library(maptpx)
library(gplots)
library(philentropy)
library(dplyr)
library(edgeR)
library(tsne)
library(qtlcharts)
library(limma)
```

## Preparing the Data

```{r}
reads <- data.frame(fread('../data/reads.txt'),row.names=1);
molecules <-  data.frame(fread('../data/molecules.txt'),row.names=1);
```


```{r}
quality_single_cells <- scan("../data/quality-single-cells.txt",
                             what = "character")
anno <- data.frame(fread('../data/annotation.txt'));
```


Remove bulk samples and include only quality single cells.

```{r}
molecules <- molecules[, grepl("bulk", colnames(molecules)) |
                         colnames(molecules) %in% quality_single_cells]
anno <- anno[anno$well == "bulk" | anno$sample_id %in% quality_single_cells, ]
stopifnot(ncol(molecules) == nrow(anno),
          colnames(molecules) == anno$sample_id)

reads <- reads[, grepl("bulk", colnames(reads)) |
                         colnames(reads) %in% quality_single_cells]
stopifnot(ncol(reads) == nrow(anno),
          colnames(reads) == anno$sample_id)
```


Remove genes with 0 count across single cells.

```{r}
expressed <- rowSums(molecules[, anno$well == "bulk"]) > 0 &
             rowSums(molecules[, anno$well != "bulk"]) > 0
molecules <- molecules[expressed, ]

expressed <- rowSums(reads[, anno$well == "bulk"]) > 0 &
             rowSums(reads[, anno$well != "bulk"]) > 0
reads <- reads[expressed, ]

molecules_single <- molecules %>% select(-contains("bulk"))
reads_single <- reads %>% select(-contains("bulk"))
```


Remove genes with max molecule numer larger than 1024

```{r}
molecules_single <- molecules_single[apply(molecules_single,1,max) < 1024,];
```


## Cell-cycle genes

Now we draw a list of marker genes that have cell cycle information.

```{r}
cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})

labs <- unique(unlist(lapply(1:5, function(k) 
                            match(cell_cycle_genes_list[[k]],
                            rownames(molecules_single)))) )
labs <- labs[!is.na(labs)]
molecules_single_cell_cycle <- molecules_single[labs,]
```


We fix the cell phase scores

```{r}
ans <-
    sapply(cell_cycle_genes_list,function(xx){
      #### create table of each phase
      reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
      #### add average expression of all genes in the phase
      combined_matrix <- rbind(reads_single_phase,average=apply(reads_single_phase,2,mean))
      #### use transpose to compute cor matrix
      cor_matrix <- cor(t(combined_matrix))
      #### take the numbers
      cor_vector <- cor_matrix[,dim(cor_matrix)[1]]
      #### restrict to correlation >= 0.3 
      reads_single_phase_restricted <- reads_single_phase[rownames(reads_single_phase) %in% names(cor_vector[cor_vector >= 0.3]),]
      #### apply normalization to reads
      norm_factors_single <- calcNormFactors(reads_single_phase_restricted, method = "TMM")
      reads_single_cpm <- cpm(reads_single_phase_restricted, log = TRUE,
                                lib.size = colSums(reads_single) * norm_factors_single)
      #### output the phase specific scores (mean of normalized expression levels in the phase)
      apply(reads_single_cpm,2,mean)
})

#### normalization function
flexible_normalization <- function(data_in,by_row=TRUE){
  if(by_row){
    row_mean <- apply(data_in,1,mean)
    row_sd   <- apply(data_in,1,sd)
    output <- data_in
    for(i in 1:dim(data_in)[1]){
      output[i,] <- (data_in[i,] - row_mean[i])/row_sd[i]
    }
  }
  #### if by column
  if(!by_row){
    col_mean <- apply(data_in,2,mean)
    col_sd   <- apply(data_in,2,sd)
    output <- data_in
    for(i in 1:dim(data_in)[2]){
      output[,i] <- (data_in[,i] - col_mean[i])/col_sd[i]
    }
  }
  output
}

#### apply the normalization function
## first normalized for each phase
ans_normed <- flexible_normalization(ans,by_row=FALSE)
## then normalized of each cell
ans_normed_normed <- flexible_normalization(ans_normed,by_row=TRUE)

cell_phase <- apply(ans_normed_normed,1,function(x) colnames(cell_cycle_genes)[which.max(x)])
assign_cell_phase <- data.frame(cell_phase)
cell_phase_vector <- as.vector(as.matrix(assign_cell_phase));
cell_phase_vector <- factor(cell_phase_vector, 
                            levels = c("G1.S", "S", "G2.M", "M", "M.G1"))
```

```{r}
counts_data <- t(molecules_single_cell_cycle)
cpm_data <- voom(counts_data)$E

individual_id <- sapply(1:length(rownames(cpm_data)), 
                        function(x) strsplit(rownames(cpm_data)[x],"[.]")[[1]][1])

batch_id <- sapply(1:length(rownames(cpm_data)), 
                   function(x) strsplit(rownames(cpm_data)[x],"[.]")[[1]][2])

individual.batch.id <- paste0(individual_id,"_", batch_id);

```


```{r}
source('../R/poisson_loglik.R')
source('../R/simplex_functions.R')
source('../R/estimation_poisson_batch.R')
source('../R/poisson_topic_loglink.R')

if(!dir.exists("../figures/poiss_batch_cellcycle")) dir.create("../figures/poiss_batch_cellcycle")


if(file.exists("../../project/rdas/poiss_randbatch_cellcycle.rda"))
{
  poiss_batch <- get(load("../../project/rdas/poiss_randbatch_cellcycle.rda"));
} else {
nclus_vec <- 2:4
Topic_clus_list <- lapply(nclus_vec, function(per_clust) {
out <- Poisson_topic.loglink(as.matrix(counts_data),per_clust,as.numeric(as.factor(batch_id)),use_squarem = FALSE,max_iter = 20)
  })

names(Topic_clus_list) <- paste0("clust_",nclus_vec)
save(Topic_clus_list, file = "../../project/rdas/poiss_randbatch_cellcycle.rda");
poiss_batch <- get(load("../../project/rdas/poiss_randbatch_cellcycle.rda"));
}

for(num in 1:length(nclus_vec))
{
obj <- StructureObj_omega(poiss_batch[[num]]$omega, samp_metadata = samp_metadata, batch_lab = batch_id,partition = rep("TRUE",dim(samp_metadata)[2]),path_struct="../figures/poiss_batch_cellcycle",control=list(cex.axis=1));
}
```

