<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Joyce Hsiao" />

<meta name="date" content="2015-12-18" />

<title>Use PCA to visualize single-cell data</title>

<script src="libs/jquery-1.11.0/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.1/css/united.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.1/shim/respond.min.js"></script>

<style type="text/css">

/* padding for bootstrap navbar */
body {
  padding-top: 50px;
  padding-bottom: 40px;
}


/* offset scroll position for anchor links (for fixed navbar)  */
.section h2 {
  padding-top: 55px;
  margin-top: -55px;
}
.section h3 {
  padding-top: 55px;
  margin-top: -55px;
}



/* don't use link color in navbar */
.dropdown-menu>li>a {
  color: black;
}

/* some padding for disqus */
#disqus_thread {
  margin-top: 45px;
}

</style>

<link rel="stylesheet" href="libs/font-awesome-4.1.0/css/font-awesome.min.css"/>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/textmate.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">


<div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ashlar</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li><a href="index.html">Home</a></li>
        <li><a href="about.html">About</a></li>
        <li><a href="license.html">License</a></li>
        <li><a href="https://github.com/jhsiao999/ashlar">GitHub</a></li>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">
<h1 class="title">Use PCA to visualize single-cell data</h1>
<h4 class="author"><em>Joyce Hsiao</em></h4>
<h4 class="date"><em>2015-12-18</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#objectives">Objectives</a></li>
<li><a href="#prepare-data">Prepare data</a></li>
<li><a href="#making-pca-plots">Making PCA plots</a></li>
<li><a href="#session-information">Session information</a></li>
</ul>
</div>

<p><strong>Last updated:</strong> 2016-01-15</p>
<p><strong>Code version:</strong> 41f37793a0adb3cc0b0c019bf40649af6d21c8dd</p>
<div id="objectives" class="section level2">
<h2>Objectives</h2>
<p>I imported a subset of the <a href="http://jdblischak.github.io/singleCellSeq/analysis">singleCellSeq</a> data and performed a principal component analysis to explore visually the similarity between batches in transcriptional expression profile. The data set includes 225 induced pluripotent stem cells (iPSCs) derived from a Yoruba cell line (NA19239) and a subsample of 1,000 genes). The cells were collected from the same cell culture and processed in three different batches.</p>
<p>I used the PCA functions created by <a href="http://jdblischak.github.io">John Blischak</a> (<em>ashlar/codes</em>) to inspect the first two major sources of variation in expression profiles. I was interested in learning: Do cells collected in different batches share similar variation? A batch</p>
</div>
<div id="prepare-data" class="section level2">
<h2>Prepare data</h2>
<p>Import gene by sample (cell) matrix. Each column consists of molecule counts of a single cell.</p>
<pre class="r"><code>molecules_filter_subsample &lt;- read.table(&quot;../data/molecules-filter-subsample.txt&quot;,
                                         header = TRUE, 
                                         stringsAsFactors = FALSE)
dim(molecules_filter_subsample)</code></pre>
<pre><code>[1] 1000  225</code></pre>
<p>Import metadata information. This annotates for every cell the individual (cell line) labels, replicate labels, etc.</p>
<pre class="r"><code>annotation_filter_subsample &lt;- read.table(&quot;../data/annotation-filter-subsample.txt&quot;,
                                          header = TRUE,
                                          stringsAsFactors = FALSE)
head(annotation_filter_subsample)</code></pre>
<pre><code>    individual replicate well      batch      sample_id
344    NA19239        r1  A01 NA19239.r1 NA19239.r1.A01
345    NA19239        r1  A03 NA19239.r1 NA19239.r1.A03
346    NA19239        r1  A05 NA19239.r1 NA19239.r1.A05
347    NA19239        r1  A06 NA19239.r1 NA19239.r1.A06
348    NA19239        r1  A07 NA19239.r1 NA19239.r1.A07
349    NA19239        r1  A08 NA19239.r1 NA19239.r1.A08</code></pre>
<pre class="r"><code>dim(annotation_filter_subsample)</code></pre>
<pre><code>[1] 225   5</code></pre>
</div>
<div id="making-pca-plots" class="section level2">
<h2>Making PCA plots</h2>
<p>Source PCA functions.</p>
<pre class="r"><code>source(&quot;../code/functions.R&quot;)</code></pre>
<p>Run PCA.</p>
<pre class="r"><code>pca_results &lt;- run_pca(as.matrix(molecules_filter_subsample))</code></pre>
<p>Visualize PCA results.</p>
<pre class="r"><code>plot_pca(pca_results$PCs,
         explained = pca_results$explained,
         metadata = annotation_filter_subsample,
         color = &quot;batch&quot;) +
  labs(title = &quot;Single-cell transcript counts&quot;)</code></pre>
<pre><code>Warning: package &#39;ggplot2&#39; was built under R version 3.2.3</code></pre>
<p><img src="figure/single-cell-pca.Rmd/unnamed-chunk-5-1.png" title="" alt="" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.2.1 (2015-06-18)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_2.0.0 knitr_1.11   

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.2      digest_0.6.8     plyr_1.8.3       grid_3.2.1      
 [5] gtable_0.1.2     formatR_1.2.1    magrittr_1.5     evaluate_0.8    
 [9] scales_0.3.0     stringi_1.0-1    rmarkdown_0.9.2  labeling_0.3    
[13] tools_3.2.1      stringr_1.0.0    munsell_0.4.2    yaml_2.1.13     
[17] colorspace_1.2-6 htmltools_0.3   </code></pre>
</div>


<!-- some extra javascript for older browsers -->
<script type="text/javascript" src="libs/polyfill.js"></script>

<script>

// manage active state of menu based on current page
$(document).ready(function () {

    // active menu
    href = window.location.pathname
    href = href.substr(href.lastIndexOf('/') + 1)
    $('a[href="' + href + '"]').parent().addClass('active');

    // manage active menu header
    if (href.startsWith('authoring_'))
      $('a[href="' + 'authoring' + '"]').parent().addClass('active');
    else if (href.endsWith('_format.html'))
      $('a[href="' + 'formats' + '"]').parent().addClass('active');
    else if (href.startsWith('developer_'))
      $('a[href="' + 'developer' + '"]').parent().addClass('active');

});

</script>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
