---
title: 'iPSC Clustering : Yoav Data'
author: "Kushal K Dey"
date: "July 10, 2015"
output: html_document
---

We direct to the webpage of J. Blischak [Webpage](http://jdblischak.github.io/singleCellSeq/analysis/index.html)

```{r packages, echo=FALSE, eval=TRUE}
suppressMessages(suppressWarnings(library(data.table)))
suppressMessages(suppressWarnings(library(maptpx)))
suppressMessages(suppressWarnings(library(gplots)))
suppressMessages(suppressWarnings(library(philentropy)))
suppressMessages(suppressWarnings(library(dplyr)))

```

We read the data on 3 individuals and 291 iPSC coming from these 3 individuals. In the sample names, there were three information clubbed (individual, batch and well). An instance was "NA19098.1.A01" where "NA19098" is the individual name, "1" is the batch label and "A01" is the well name.

```{r data_loading, echo=TRUE, eval=TRUE}

#setwd('/Users/kushal/Documents/Matthew Stephens Project/counts_clustering/Single_Cell_seq_Blischak/src/');
reads <- data.frame(fread('../data/reads.txt'),row.names=1);
molecules <-  data.frame(fread('../data/molecules.txt'),row.names=1);


quality_single_cells <- scan("../data/quality-single-cells.txt",
                             what = "character")

anno <- data.frame(fread('../data/annotation.txt'));


molecules <- molecules[, grepl("bulk", colnames(molecules)) |
                         colnames(molecules) %in% quality_single_cells]
anno <- anno[anno$well == "bulk" | anno$sample_id %in% quality_single_cells, ]
stopifnot(ncol(molecules) == nrow(anno),
          colnames(molecules) == anno$sample_id)

reads <- reads[, grepl("bulk", colnames(reads)) |
                         colnames(reads) %in% quality_single_cells]
stopifnot(ncol(reads) == nrow(anno),
          colnames(reads) == anno$sample_id)


## Filtering out genes which have zero counts in single cells/ bulk

expressed <- rowSums(molecules[, anno$well == "bulk"]) > 0 &
             rowSums(molecules[, anno$well != "bulk"]) > 0
molecules <- molecules[expressed, ]
dim(molecules)

expressed <- rowSums(reads[, anno$well == "bulk"]) > 0 &
             rowSums(reads[, anno$well != "bulk"]) > 0
reads <- reads[expressed, ]
dim(reads)

molecules_single <- molecules %>% select(-contains("bulk"))
reads_single <- reads %>% select(-contains("bulk"))

molecules_single <- molecules_single[apply(molecules_single,1,max) < 1024,]


```



What is the difference between molecule data and reads data?....The counts for the molecule data seem to be less compared to the reads. The number of reads coming from a sample on an average for the reads and the molecules data is as follows 


## Cell Cycle Phases

```{r echo=TRUE, eval=TRUE}

cell_cycle_genes <- read.table("../data/cellcyclegenes.txt", header = TRUE, sep="\t")

## create 5 lists of 5 phases (de-level and then remove "")
cell_cycle_genes_list <- lapply(1:5,function(x){
  temp <- as.character(cell_cycle_genes[,x])
  temp[temp!=""]
})

phase_reads_counts <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  reads_single_phase <- reads_single[rownames(reads_single) %in% unlist(xx) ,]
  combined_phase <-apply(reads_single_phase,2,sum);
  })

phase_molecules_counts <-
sapply(cell_cycle_genes_list,function(xx){
  #### create table of each phase
  molecules_single_phase <- molecules_single[rownames(molecules_single) %in% unlist(xx) ,]
  combined_phase <-apply(molecules_single_phase,2,sum);
  })
```


### Topic model (individual effects) on cell cycles data

```{r echo=FALSE, eval=TRUE}

Topic_Clus_molecules <- topics(phase_molecules_counts,K=7,tol=0.001);

Topic_Clus_reads <- topics(phase_reads_counts,K=7,tol=0.001);

docweights_molecules <- Topic_Clus_molecules$omega;
K=7

color=c("red","blue","cornflowerblue","black","cyan","darkblue",
        "brown4","burlywood","darkgoldenrod1","darkgray","deepskyblue","darkkhaki",
        "firebrick","darkorchid","hotpink","green","magenta","yellow", "azure1","azure4");

subgroup_id =rownames(docweights_molecules);

individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
#cell_subtype = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));

individual_effect_ordered = individual_effect [order(individual_effect)];
docweights_molecules_ordered = docweights_molecules[order(individual_effect),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(individual_effect_ordered), individual_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(individual_effect_ordered),las=2);


```

```{r echo=FALSE, eval=TRUE}

docweights_reads <- Topic_Clus_reads$omega;
K=7

color=c("red","blue","cornflowerblue","black","cyan","darkblue",
        "brown4","burlywood","darkgoldenrod1","darkgray","deepskyblue","darkkhaki",
        "firebrick","darkorchid","hotpink","green","magenta","yellow", "azure1","azure4");

subgroup_id =rownames(docweights_reads);

individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
#cell_subtype = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));

individual_effect_ordered = individual_effect [order(individual_effect)];
docweights_reads_ordered = docweights_reads[order(individual_effect),];

barplot(t(docweights_reads_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(individual_effect_ordered), individual_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_reads_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(individual_effect_ordered),las=2);



```


### Topic model (batch effects) on cell cycles data

```{r echo=FALSE, eval=TRUE}

batch_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));

batch_effect_ordered = batch_effect [order(batch_effect)];
docweights_molecules_ordered = docweights_molecules[order(batch_effect),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(batch_effect_ordered), batch_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(batch_effect_ordered),las=2);

```

```{r echo=FALSE, eval=TRUE}

batch_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));

batch_effect_ordered = batch_effect [order(batch_effect)];
docweights_reads_ordered = docweights_reads[order(batch_effect),];

barplot(t(docweights_reads_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(batch_effect_ordered), batch_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_reads_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(batch_effect_ordered),las=2);

```

### Topic model (well effects) on cell cycles data

```{r, echo=FALSE, eval=TRUE}

well_effect_mod = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));

well_effect_ordered = well_effect_mod [order(well_effect_mod)];
docweights_molecules_ordered = docweights_molecules[order(well_effect_mod),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(well_effect_ordered), well_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(well_effect_ordered),las=2);
```

```{r, echo=FALSE, eval=TRUE}

well_effect_mod = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));

well_effect_ordered = well_effect_mod [order(well_effect_mod)];
docweights_reads_ordered = docweights_reads[order(well_effect_mod),];

barplot(t(docweights_reads_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(well_effect_ordered), well_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_reads_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(well_effect_ordered),las=2);
```



```{r reads_compare, echo=TRUE, eval=TRUE}

mean(colSums(reads))
mean(colSums(molecules))

```



##  Topic model on the molecules data

```{r topic_model_molecules, echo=FALSE, eval=TRUE}


## run topic model on the counts data 

K=7;
# Topic_Clus <- topics(t(counts_data),K=7,tol=0.001);

#write.table(Topic_Clus$omega, paste0('/Users/kushal/Documents/Matthew Stephens Project/counts_clustering/Single_Cell_seq_Blischak/data_derived/topics_omega_molecules_clus_',K));
#write.table(Topic_Clus$theta, paste0('/Users/kushal/Documents/Matthew Stephens Project/counts_clustering/Single_Cell_seq_Blischak/data_derived/topics_theta_molecules_clus_',K));


docweights_molecules <- read.table(paste0('../data_derived/topics_omega_molecules_clus_',K));

```

Building the Structure Plot for the topic weights for the molecules data based on the individual specific effects (potential batch effects??)

```{r, echo=FALSE, eval=TRUE}

####  Structure plot

color=c("red","blue","cornflowerblue","black","cyan","darkblue",
        "brown4","burlywood","darkgoldenrod1","darkgray","deepskyblue","darkkhaki",
        "firebrick","darkorchid","hotpink","green","magenta","yellow", "azure1","azure4");

#barplot(t(docweights_molecules),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

subgroup_id =rownames(docweights_molecules);

individual_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[1]));
#cell_subtype = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));

individual_effect_ordered = individual_effect [order(individual_effect)];
docweights_molecules_ordered = docweights_molecules[order(individual_effect),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(individual_effect_ordered), individual_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(individual_effect_ordered),las=2);
```

Building the Structure Plot for the topic weights for the molecules data based on the batches (further clarification needed what these are)...look up [JBlischak page](http://jdblischak.github.io/singleCellSeq/analysis/sum-counts-per-sample.html)

```{r, echo=FALSE, eval=TRUE}

batch_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[2]));

batch_effect_ordered = batch_effect [order(batch_effect)];
docweights_molecules_ordered = docweights_molecules[order(batch_effect),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(batch_effect_ordered), batch_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(batch_effect_ordered),las=2);
```

Building the Structure Plot for the topic weights for the molecules data based on the wells

```{r, echo=FALSE, eval=TRUE}

well_effect = unlist(lapply(strsplit(subgroup_id, "\\."), function(x) x[3]));
well_effect_mod=character(length(well_effect));
well_effect_mod[which(well_effect=="bulk")]="bulk";
well_effect_mod[-(which(well_effect=="bulk"))] = substring(well_effect[-(which(well_effect=="bulk"))],1,1)

well_effect_ordered = well_effect_mod [order(well_effect_mod)];
docweights_molecules_ordered = docweights_molecules[order(well_effect_mod),];

barplot(t(docweights_molecules_ordered),col=color[1:K],axisnames=F,space=0,border=NA,main=paste("No. of clusters=",K),las=1,ylim=c(0,1),cex.axis=1.5,cex.main=1.4)

labels = match(unique(well_effect_ordered), well_effect_ordered);
abline(v=labels)

labels_low=labels;
labels_up=c(labels[2:length(labels)],dim(docweights_molecules_ordered)[1]);
mid_point=labels_low +0.5*(labels_up-labels_low);

axis(1,at=mid_point, unique(well_effect_ordered),las=2);
```






